mmv1/products/compute/TargetHttpsProxy.yaml

```diff
@@ -115,19 +126,34 @@ properties:
     update_url: 'projects/{{project}}/global/targetHttpsProxies/{{name}}/setQuicOverride'
     default_value: :NONE
     custom_flatten: 'templates/terraform/custom_flatten/default_if_empty.erb'
+  - !ruby/object:Api::Type::Array
+    name: 'certificateManagerCertificates'
+    description: |
+      A list of Certificate Manager certificate URLs that are used to authenticate
+      connections between users and the load balancer. At least one resource must be specified.
+      Accepted format is `//certificatemanager.googleapis.com/projects/{project}/locations/{location}/certificates/{resourceName}` or just the self_link projects/{project}/locations/{location}/certificates/{resourceName}
+    update_verb: :POST
+    update_url: 'projects/{{project}}/targetHttpsProxies/{{name}}/setSslCertificates'
+    item_type: Api::Type::String
+    custom_expand: 'templates/terraform/custom_expand/certificate_manager_certificate_construct_full_url.go.erb'
+    diff_suppress_func: 'tpgresource.CompareResourceNames'
```

Comment:
Is this necessary? I'm not sure I understand how it is being used

Reply:
Do you mean the `diff_suppress_func` or something else?  

The API expects each item to be in full URL format (e.g. `//certificatemanager.googleapis.com/projects/../locations/...`. 
However, I allowed the users to only provide the `Id`, so customers can just provide each reference on the format of `projects/../locations/../certficates/..`, and I construct the full URL before sending the API request using the custom_expand function.
So, we end up having the following: 
- **user input** -> `projects/../locations/../certificates/..`
-  **API response**  -> `//certificatemanager.googleapis.com/projects/../locations/.../..certificates/..`

When I run the test, TF triggers an update request(because the field value has changed) and that causes the test to fail. 

Is there a better way to handle this situation?

Reply:
One more question, Are there any issues with allowing the user to provide the `id` and then convert it to a full URL before sending the request? 

According to rileykarson's comment here https://github.com/GoogleCloudPlatform/magic-modules/pull/8941#issuecomment-1747414818, it's not encouraged to let the user enter the name of the resource unless the API allows doing this, I wonder if it's also the same with letting the user provide an `id` if the API is expecting a full URL format.

Reply:
Yes, I'm talking about the `diff_suppress_func`. I understand that you are allowing the user to provide an id (which I believe is fine here), and that needs to be converted before being sent to the API, but isn't this all being handled by the expand/encoder/decoder?

I may just be missing something, but I'm wondering what caused the test failure. And for context, I'm partly asking about the diff suppress because I believe the implementation doesn't consider all parts of the id (like location), and that may cause problems.

Reply:
1-  The encoder and the decoder parts are responsible for merging or splitting the fields `sslCertificates` and `certificate_manager_certificates` before constructing the API request and after receiving the response respectively. They have nothing to do with the format of the input provided by the user. 

2- The custom expand is responsible for making sure that any certificate manager certificate resource is sent in the API request in the format of `//certificatemanager.googleapis.com/projects/../locations/.../..certificates/..`. So, if the user uses the `id` to reference the certificate, the custom expand will convert it before sending the API request. 

The response that we receive from the API will contain the certificates in the full URL format (`//certificatemanager.googleapis.com/projects/../locations/.../..certificates/..`). Hence if the user provides only the `id`, there will be a difference between the input provided by the user and the response from the API. 

**Error Message:** 
```
2023-10-06T18:46:25.436Z [ERROR] sdk.helper_resource: Unexpected error:
  error=
  | After applying this test step, the plan was not empty.
  | stdout:
  | 
  | 
  | Terraform used the selected providers to generate the following execution
  | plan. Resource actions are indicated with the following symbols:
  |   ~ update in-place
  | 
  | Terraform will perform the following actions:
  | 
  |   # google_compute_target_https_proxy.default will be updated in-place
  |   ~ resource "google_compute_target_https_proxy" "default" {
  |       ~ certificate_manager_certificates = [
  |           - "//certificatemanager.googleapis.com/projects/hamzahassan/locations/global/certificates/tf-test-my-certificate4di3je7hnx",
  |           + "projects/hamzahassan/locations/global/certificates/tf-test-my-certificate4di3je7hnx",
  |         ]
  |         id                               = "projects/hamzahassan/global/targetHttpsProxies/tf-test-target-http-proxy4di3je7hnx"
  |         name                             = "tf-test-target-http-proxy4di3je7hnx"
  |         # (8 unchanged attributes hidden)
  |     }
  | 
  | Plan: 0 to add, 1 to change, 0 to destroy.
```
* This error only happens if the user provides `id` to reference a certificate resource

Reply:
Got it, that all makes sense, thanks for clarifying. Despite my concern with the other parts of the id diffing (like location and project), it appears this same function is used in many other fields with similar structure, so seems good to me.